core/internal/leet/reader.go

```diff
@@ -0,0 +1,298 @@
+				Project:     rec.Run.Project,
+				Config:      rec.Run.Config,
+			}
+		}
+	case *spb.Record_History:
+		if rec.History != nil {
+			return ParseHistory(rec.History)
+		}
+	case *spb.Record_Stats:
+		if rec.Stats != nil {
+			return ParseStats(rec.Stats)
+		}
+	case *spb.Record_Summary:
+		if rec.Summary != nil {
+			return SummaryMsg{Summary: rec.Summary}
+		}
+	case *spb.Record_Environment:
+		if rec.Environment != nil {
+			return SystemInfoMsg{Record: rec.Environment}
+		}
+	case *spb.Record_Exit:
+		if rec.Exit != nil {
+			return FileCompleteMsg{ExitCode: rec.Exit.ExitCode}
+		}
+	}
+	return nil
+}
+
+// ParseHistory extracts metrics from a history record.
+func ParseHistory(history *spb.HistoryRecord) tea.Msg {
+	if history == nil {
+		return nil
+	}
+
+	metrics := make(map[string]float64)
+	var step int
+
+	for _, item := range history.Item {
+		if item == nil {
+			continue
+		}
+
+		key := strings.Join(item.NestedKey, ".")
+		if key == "_step" {
+			if val, err := strconv.Atoi(strings.Trim(item.ValueJson, `"`)); err == nil {
+				step = val
+			}
+			continue
+		}
+
+		if strings.HasPrefix(key, "_") {
+			continue
+		}
+
+		if value, err := strconv.ParseFloat(strings.Trim(item.ValueJson, `"`), 64); err == nil {
+			metrics[key] = value
+		}
+	}
+
+	if len(metrics) > 0 {
+		return HistoryMsg{Metrics: metrics, Step: step}
+	}
+	return nil
+}
+
+// ParseStats extracts metrics from a stats record.
+func ParseStats(stats *spb.StatsRecord) tea.Msg {
+	if stats == nil {
+		return nil
+	}
+
+	metrics := make(map[string]float64)
+	var timestamp int64
+
+	if stats.Timestamp != nil {
+		timestamp = stats.Timestamp.Seconds
+	}
+
+	for _, item := range stats.Item {
+		if item == nil {
+			continue
+		}
+
+		if value, err := strconv.ParseFloat(strings.Trim(item.ValueJson, `"`), 64); err == nil {
+			metrics[item.Key] = value
+		}
+	}
+
+	if len(metrics) > 0 {
+		return StatsMsg{Timestamp: timestamp, Metrics: metrics}
+	}
+	return nil
+}
+
+// Close closes the reader.
+func (r *WandbReader) Close() {
+	if r.store != nil {
+		r.store.Close()
+	}
+}
+
+// InitializeReader creates a command to initialize the wandb reader.
+func InitializeReader(runPath string, logger *observability.CoreLogger) tea.Cmd {
+	return func() tea.Msg {
+		reader, err := NewWandbReader(runPath, logger)
+		if err != nil {
+			return ErrorMsg{Err: err}
+		}
+		return InitMsg{Reader: reader}
+	}
+}
+
+// ReadAllRecordsChunked returns a command to read records in chunks for progressive loading.
+func ReadAllRecordsChunked(reader *WandbReader) tea.Cmd {
+	if reader == nil {
+		// No reader available; no-op to keep Bubble Tea flow consistent.
+		return func() tea.Msg { return nil }
+	}
+	return reader.ReadAllRecordsChunked()
+}
+
+// ReadAvailableRecords reads new records for live monitoring.
+func ReadAvailableRecords(reader *WandbReader) tea.Cmd {
+	// No reader? Nothing to do.
+	if reader == nil {
+		return func() tea.Msg { return nil }
+	}
```

Comment:
For this also, I recommend removing the nil check and unwrapping the nested func. The implementation becomes simpler and the callsites more explicitly signal that no-op is a possibility.